### PR TITLE
Reland caching allocator for avoiding reallocations (#1519)

### PR DESCRIPTION
**Pull request by rsuderman**
We can avoid reallocating at the hip level by cacheing allocations between invocations. This is particularly useful for decode which sees sets of 32 decode steps with the same dimensions and dtypes.

Previous version failed to remove reused buffers from the list of cached tensors. This resulted in potential buffer reuse and could cause a crash when attempting to find that cached buffer in the future.

Caching limit was also changed to rebuild the cache / tables rather than mutate them. This avoids unintended effects of the `del` operator.

### CODE REVIEW

File: shortfin/python/shortfin_apps/llm/components/batcher.py [hunk 1]
```diff
@@ -360,6 +374,8 @@ async def run(self):
                 req_count=req_count, device0=device0, buffers=(logits, indices)
             )
 
+            [arg.release() for arg in args]
```
[the remark this answers is not in the document]
Comment (stbaione) [reply shortfin/python/shortfin_apps/llm/components/batcher.py:377]: We alias the host side logits array with `view` when we set the `result_logits` of an `LlmInferenceExecRequest`. This means that it'd only be safe to re-use a cache logits array if all requests within a batch were finished. Would be better to make this change in another PR

[the remark this answers is not in the document]
Comment (rsuderman) [reply shortfin/python/shortfin_apps/llm/components/batcher.py:377]: There is some complexity due to aliasing the memory. If we perform our views then copy we can reuse the host buffers. Once `topk` work is in we can check whether copy and cache would work (but that may just mean more allocations).
